templates/terraform/constants/snapshot.erb

```diff
@@ -0,0 +1,133 @@
+	// and then has another different value, so we ForceNew.
+	// We need to handle _EVERY_ ForceNew case in this diff
+	if oldConvenience != "" && newConvenience != "" && oldConvenience != newConvenience {
+		return diff.ForceNew("snapshot_encryption_key_raw")
+	}
+
+	if oldNewField != "" && newNewField != "" && oldNewField != newNewField {
+		return diff.ForceNew("snapshot_encryption_key.0.raw_key")
+	}
+
+	// Our resource isn't using either field, then uses one;
+	// ForceNew on whichever one is now using it.
+	if (oldConvenience == "" && oldNewField == "" && newConvenience != "") || (oldConvenience == "" && oldNewField == "" && newNewField != "") {
+		if oldConvenience == "" && newConvenience != "" {
+			return diff.ForceNew("snapshot_encryption_key_raw")
+		} else {
+			return diff.ForceNew("snapshot_encryption_key.0.raw_key")
+		}
+	}
+
+	// convenience no longer used
+	if oldConvenience != "" && newConvenience == "" {
+		if newNewField == "" {
+			// convenience is being nulled, and the new field is empty as well
+			// we've stopped using the field altogether
+			return diff.ForceNew("snapshot_encryption_key_raw")
+		} else if oldConvenience != newNewField {
+			// convenience is being nulled, and the new field has a new value
+			// so we ForceNew on either field
+			return diff.ForceNew("snapshot_encryption_key_raw")
+		} else {
+			// If we reach it here, we're using the same value in the new field as we had in the convenience field
+		}
+	}
+
+	// new no longer used
+	// note that it will remain _set_ because of how Computed fields work
+	// unset fields will have their values kept in state as a non-zero value
+	if oldNewField != "" && newNewField == "" {
+		if newConvenience == "" {
+			// new field is being nulled, and the convenience field is empty as well
+			// we've stopped using the field altogether
+			return diff.ForceNew("snapshot_encryption_key.0.raw_key")
+		} else if oldNewField != newConvenience {
+			// new is being nulled, and the convenience field has a new value
+			// so we ForceNew on either field
+
+			// This stops a really opaque diffs don't match during apply error. Without this, wee see
+			// a diff from the old state -> new state with a ForceNew at plan time (as expected!)
+			// But during apply time the entire nested object is nil in old state unexpectedly.
+			// So we just force the diff to match more by nilling it here, which is unclear why it
+			// works, and probably a worse UX with some real ugly diff, but also makes the tests pass.
+			// Computed nested fields are hard.
+			err := diff.SetNew("snapshot_encryption_key", nil)
+			if err != nil {
+				return err
+			}
+
+			return diff.ForceNew("snapshot_encryption_key.0.raw_key")
+		} else {
+			// If we reach it here, we're using the same value in the convenience field as we had in the new field
+		}
+	}
+
+	return nil
+}
+
+func customDiffComputeSnapshotSourceDiskEncryptionKeys(diff *schema.ResourceDiff, meta interface{}) error {
```

Comment:
This looks the same as above - can it be parameterized?  If so, it would help us migrate convenience fields in the future.

Reply:
So, these are mostly the same but different enough that I don't think that's worth doing. `snapshotEncryptionKey` is way more fiddly than `sourceDiskEncryptionKey` and I expect the next similar field will have just as many differences from either of them.

Reply:
Great!  LGTM.